### DISCUSSION TITLE
Value mapping

### DISCUSSION THREAD

**Discussion by mbretter** [react-hook-form/react-hook-form#1935] [Q&A] [answered]
Hi,

is there any possibility to map the value before it will be set into the input component?

Some api returns boolean values as integer 0/1, when setting these values into a checkbox, it will not work, because integers are not a valid value for the checked property.

The other use case would be implicitly converting undefined/null into an empty string for TextBoxes.

Using the onChange event of the controller, the value could be converted back, but how can this be done the other way?

Here's a code snippet (MUI checkbox)
```
    const {control, errors} = useFormContext()

   return <FormControl fullWidth={fullWidth} error={error}>
        <Controller control={control}
                    name={name}
                    onChange={([, v]) => {
                        return v ? 1 : 0
                    }}
                    as={<FormControlLabel
                        label={label}
                        control={<Checkbox color="primary" {...checkboxProps}/>}
                    />}
                    defaultValue={false}
        />
        <FormHelperText>{helperText}</FormHelperText>
    </FormControl>
```

thx, cheers

**Comment by bluebill1049** [comment, marked as answer]
i think you will have to build a wrapper around MUI checkbox then, which will accept 1 and 0 as true and false.

**Reply by mbretter** [reply]
could be a solution, but the possibility to map the value would be a neat thing ;)
Something like:
```
const {control, errors, getValues} = useFormContext()
let {value} = getValues(name)

   value = Boolean(value)
   return <FormControl fullWidth={fullWidth} error={error}>
        <Controller control={control}
                    name={name}
                    value={value}
                    onChange={([, v]) => {
                        return v ? 1 : 0
                    }}
                    as={<FormControlLabel
                        label={label}
                        control={<Checkbox color="primary" {...checkboxProps}/>}
                    />}
                    defaultValue={false}
        />
        <FormHelperText>{helperText}</FormHelperText>
    </FormControl>
```

**Reply by bluebill1049** [reply]
yea, it's always a tradeoff as well, I prefer to Component as primitive as possible without blowing up the APIs/Props.

**Reply by mbretter** [reply]
The simple int to bool conversion is only one usecase, some of the apis are returning strings like "active", building a component for every variant and UI framework would be a bit expensive. 
The solution above could help when dealing with textfields and null/undefined values too, this happens a lot when working with noSQL databases, where some fields are not present (soft migration of data).
I moved over from formik, because react-hook-form does the job much better for us, but this was the only feature we missed :worried:.

**Reply by bluebill1049** [reply]
Valid points above, is there a chance to build a generic wrapper for those value mapping?

**Reply by bluebill1049** [reply]
feel free to suggest: if you have any purposed API or idea?

**Reply by mbretter** [reply]
maybe by using callbacks, the MUI wrapped component:
```JS
const FormField = ({as, mapValue, ...rest}) => {
    const {control} = useFormContext()

    return <Controller 
                       control={control}
                       mapValue={mapValue)
                       as={as || MUITextField}
                       {...rest}
    />
}
FormField.propTypes = {
    mapValue: PropTypes.func,
    as: PropTypes.element
}
```

usage in the form:
```JS
<FormField name="name" label="Name" required fullWidth mapValue={v => v === undefined ? '' : v}/>
```

**Reply by bluebill1049** [reply]
probably shouldn't be at the controller level.

**Reply by mbretter** [reply]
hmm, without knowing the internals of rhf, I could imagine to pass an additional geter option to useForm, by passing name/value and also the ref to the field, to be able to write generic mappings for certain types of input elements.

```JS
    const methods = useForm({
        defaultValues: data,
        getValue: (name, value, fieldRef) => {
            if (value === undefined || value === null) return '' // map undefined/null to empty string
            if (name === 'status') return value === 'active' // map a silly status field value to bool

            return value
        }
    })
```

**Comment by dandv** [comment]
@mbretter: sounds just like my earlier question about [transforming the value returned by a controller](https://github.com/react-hook-form/react-hook-form/discussions/1717).
